### PR TITLE
Move boxset detection logic to CMusicDatabase

### PR DESCRIPTION
**Pull request by the-black-eagle**

### CODE REVIEW

**Comment by DaveTBlake** [comment]
OK, on it @the-black-eagle

File: xbmc/music/MusicDatabase.cpp [hunk 1]
```diff
@@ -502,7 +502,17 @@ bool CMusicDatabase::AddAlbum(CAlbum& album, int idSource)
       PrepareSQL("UPDATE album SET iDisctotal = (SELECT COUNT(DISTINCT iTrack >> 16) FROM song "
                  "WHERE song.idAlbum = album.idAlbum) WHERE idAlbum = %i",
                  album.idAlbum));
-
+  std::string strSQL;
+  strSQL = PrepareSQL(
+      "SELECT COUNT(DISTINCT strDiscSubtitle) FROM song WHERE song.idAlbum = %i", album.idAlbum);
+  int numTitles = strtol(GetSingleValue(strSQL).c_str(), nullptr, 10);
```
Comment (DaveTBlake) [inline xbmc/music/MusicDatabase.cpp]: `static_cast<int>`

File: xbmc/music/MusicDatabase.cpp [hunk 2]
```diff
@@ -502,7 +502,17 @@ bool CMusicDatabase::AddAlbum(CAlbum& album, int idSource)
...
+  if (!album.bBoxedSet) // not already a boxset via tag
+    if (!album.bCompilation) // and not a compilation album
+      if (numTitles >= 3) // 3 distinct titles already so this must be a boxset
```
Comment (DaveTBlake) [inline xbmc/music/MusicDatabase.cpp]: Done like this needs only one if statement 
`If (!album.bBoxedSet && !album.bCompilation && numTitles >= 3)`
But you could move the `SELECT COUNT(DISTINCT strDiscSubtitle) ... ` query inside a check that `!album.bBoxedSet && !album.bCompilation` -  only do query when necessary

File: xbmc/music/MusicDatabase.cpp [hunk 3]
```diff
@@ -502,7 +502,17 @@ bool CMusicDatabase::AddAlbum(CAlbum& album, int idSource)
       PrepareSQL("UPDATE album SET iDisctotal = (SELECT COUNT(DISTINCT iTrack >> 16) FROM song "
                  "WHERE song.idAlbum = album.idAlbum) WHERE idAlbum = %i",
                  album.idAlbum));
-
```
Comment (DaveTBlake) [inline xbmc/music/MusicDatabase.cpp:505]: // if has at least 3 distinct disc subtitles then make album a boxset
or something like that

File: xbmc/music/MusicDatabase.cpp [hunk 4]
```diff
@@ -612,6 +622,12 @@ int CMusicDatabase::AddSong(const int idAlbum,
     if (m_pDS->num_rows() == 0)
     {
       m_pDS->close();
+      bool isBoxset = IsAlbumBoxset(idAlbum);
+      if (isBoxset && strDiscSubtitle.empty()) // boxset tag set
+      { // create a dummy title
+        int discno = iTrack >> 16;
+        strDiscSubtitle = StringUtils::Format("%s %i", g_localizeStrings.Get(427), discno);
+      }
```
Comment (DaveTBlake) [inline xbmc/music/MusicDatabase.cpp]: Comments could explain better,  "dummy" sounds like a hack (sorry, I'll be a white space in commments monitor next)
```
    // All songs of a boxset must have a disc subtitle, use "Disc N" when not given in tags
    bool isBoxset = IsAlbumBoxset(idAlbum);
    if (isBoxset && strDiscSubtitle.empty())
    {
      int discno = iTrack >> 16;
      strDiscSubtitle = StringUtils::Format("%s %i", g_localizeStrings.Get(427), discno);
    }
```

File: xbmc/music/MusicDatabase.cpp [hunk 5]
```diff
@@ -855,7 +871,8 @@ int CMusicDatabase::AddAlbum(const std::string& strAlbum, const std::string& str
                           strArtist.c_str(),
                           strAlbum.c_str());
     m_pDS->query(strSQL);
-
+    if (strType.find("boxset") != std::string::npos) //boxset tag set
```
Comment (DaveTBlake) [inline xbmc/music/MusicDatabase.cpp]: // Boxset flagged in album type

**Review by DaveTBlake — COMMENTED** [review]
Cool, I like this implementation :)
Tidy-up and push to master and I'll get a test build off

BTW if you want my attention you can always just ping me by a mention of git

**Comment by DaveTBlake** [comment]
My brain just burped out the question: what happens if the user changes an album into a boxed set via NFO file (or JSON eventually)?

Processing in UpdateAlbum I suspect.... ensure songs have disc titles, stop for disc count = 1. 
This can be taken up when you do the JSON bit.

File: xbmc/music/MusicDatabase.cpp (hunk 2 above)
Comment (the-black-eagle) [reply xbmc/music/MusicDatabase.cpp]: Good idea, re-written.

File: xbmc/music/MusicDatabase.cpp [hunk 6]
```diff
@@ -502,7 +502,18 @@ bool CMusicDatabase::AddAlbum(CAlbum& album, int idSource)
...
+  if (!album.bBoxedSet && !album.bCompilation)
+  {
+    std::string strSQL;
+    strSQL = PrepareSQL("SELECT COUNT(DISTINCT strDiscSubtitle) FROM song WHERE song.idAlbum = %i",
+                        album.idAlbum);
+    int numTitles = static_cast<int>(strtol(GetSingleValue(strSQL).c_str(), nullptr, 10));
+    if (numTitles >=3) // 3 or more distinct disc titles so this is a boxset album
+    {
+    strSQL = PrepareSQL("UPDATE album SET bBoxedSet=1 WHERE album.idAlbum=%i", album.idAlbum);
+    m_pDS->exec(strSQL);
```
Comment (DaveTBlake) [inline xbmc/music/MusicDatabase.cpp]: Indentation (save the "bot" on master)

And I would still rather see the comment labelling this entire section of code than inline after the 2nd if

File: xbmc/music/MusicDatabase.cpp [hunk 7]
```diff
@@ -612,6 +623,12 @@ int CMusicDatabase::AddSong(const int idAlbum,
     if (m_pDS->num_rows() == 0)
     {
       m_pDS->close();
+      bool isBoxset = IsAlbumBoxset(idAlbum);
+      if (isBoxset && strDiscSubtitle.empty()) // boxset tag set but no disc subtitle in tags
+      { // As all discs in a boxset have to have a title, generate one in the form of 'Disc N'
```
Comment (DaveTBlake) [inline xbmc/music/MusicDatabase.cpp]: Replace 2 comments with a single one above bool declaration, again labelling section of code

**Review by DaveTBlake — COMMENTED** [review]
Small things - less haste more speed!

File: xbmc/music/MusicDatabase.cpp (hunk 6 above)
Comment (the-black-eagle) [reply xbmc/music/MusicDatabase.cpp]: Missed the indentation, thanks!

Regarding the comment, is this better above the if (!album.bBoxedSet && .....)
`  /* If an album is not already marked as a boxset or compilation, check how many distinct
     disc titles we already have.  If there are at least three, then the album is a boxset so mark
     it as such in the db. This allows for boxsets spread over multiple directories (CD1, CD2 etc)
  */`
The closing comment tag `*/` is not on the same line as it is above, it's on its own line.

Comment (DaveTBlake) [reply xbmc/music/MusicDatabase.cpp]: Humm... too verbose, find a happy medium (God I'm in a picky mode tonight, fetch me a beer!).

 // Set a non-compilation album with at least 3 distinct disc titles as a boxset

File: xbmc/music/MusicDatabase.cpp (hunk 7 above)
Comment (DaveTBlake) [reply xbmc/music/MusicDatabase.cpp]:
```
// As all discs in a boxset have to have a title, generate one in the form of 'Disc N
bool isBoxset = IsAlbumBoxset(idAlbum);
if (isBoxset && strDiscSubtitle.empty())
{ 
```

Comment (the-black-eagle) [reply xbmc/music/MusicDatabase.cpp]: `      // See if the current album is flagged as a boxset.  If it is, as boxset discs have to have a`
`      // title, if there is no title from the tags then generate our own in the form of 'Disc N'`
`      bool isBoxset = IsAlbumBoxset(idAlbum);`

Doesn't fit on here, but is that better ?  Before I commit it ?

File: xbmc/music/MusicDatabase.cpp (hunk 6 above)
Comment (the-black-eagle) [reply xbmc/music/MusicDatabase.cpp]: Grrr, who would have thought comments would be so difficult to get right :)

**Review by DaveTBlake — APPROVED** [review]
Go!!!

File: xbmc/music/MusicDatabase.cpp (hunk 6 above)
Comment (DaveTBlake) [reply xbmc/music/MusicDatabase.cpp]: Silly isn't it, but the quality work deserves it :)

**Comment by the-black-eagle** [comment]
Done, squashed on other branch back to the 1 commit against master.

You can go have that beer now !! :)

**Comment by DaveTBlake** [comment]
I kicked off a jenkins build first :) 
https://jenkins.kodi.tv/view/Helpers/job/BuildMulti-PR-Manually/304/
Not sure if you can see that, but it should result in installation files for various platforms on http://mirrors.kodi.tv/test-builds/ etc.

**Comment by the-black-eagle** [comment]
Yeah, I can see it fine. Thanks for the heads-up !

**Pull request merged**